compiler/rustc_const_eval/src/check_consts/check.rs

```diff
@@ -695,8 +695,7 @@ impl<'tcx> Visitor<'tcx> for Checker<'_, 'tcx> {
                     if trait_is_const {
                         // Trait calls are always conditionally-const.
                         self.check_op(ops::ConditionallyConstCall { callee, args: fn_args });
-                        // FIXME(const_trait_impl): do a more fine-grained check whether this
-                        // particular trait can be const-stably called.
+                        self.tcx.enforce_trait_const_stability(trait_did, *fn_span);
```

Comment:
This seems like it cannot work properly, since `check_op` must be used to check the feature gates. That is crucial to enforce recursive const stability rules.

Reply:
I've implemented recursive const stability enforcement in `rustc_middle` because it is needed to enforce for const stable/unstable traits as well and MIR does not have access to those.

Reply:
I was hoping the MIR const check could just ask "which trait is this function on", "give me that trait's stability", and then proceed as before... but @compiler-errors already mentioned in the past that they wanted to move the stability check so maybe this cannot be avoided.

Would it be possible to split the refactor of moving the check to rustc_middle into a separate PR? Reviewing that refactor together with the change that adds new checks for const traits sounds like a nightmare.

Reply:
traits don't necessarily have functions so mir checks could be skipping those. this only does the const traits checks and is not a full refactor of the current mir check to hir, so I don't think there are things that can be extracted from this.

Reply:
Ah, so we have the logic for this spread across two modules now? Hm, that's not great either.^^ I hope someone will clean this up in the future...

Reply:
the logic that I have added should be as deduplicated as possible, as there are functions exposed in rustc_middle to allow reusing